react/AppIcon/test/AppIcon.spec.js

```diff
@@ -0,0 +1,108 @@
+'use strict'
+/* eslint-env jest */
+
+import React from 'react'
+import { configure, shallow } from 'enzyme'
+import Adapter from 'enzyme-adapter-react-15'
+
+import AppIcon from '../'
+
+configure({ adapter: new Adapter() })
+
+describe('AppIcon component', () => {
+  const app = {
+    links: {
+      icon: 'apps/test/icon'
+    }
+  }
+
+  const context = {
+    client: {
+      link: {
+        client: {
+          fetch: jest.fn()
+        }
+      }
+    }
+  }
+
+  const iconSvg =
+    '<svg viewBox="0 0 16 16" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"> \
+      <defs> \
+          <path d="M1,11.0086308 L1,5.50051502 C1,4.95127159 1.3905874,4.72776572 1.87240137,5.00308799 L7.12759863,8.00605786 C7.60359323,8.27805477 8,8.95047139 8,9.50051502 L8,15.0086308 C8,15.5578743 7.6094126,15.7813801 7.12759863,15.5060579 L1.87240137,12.503088 C1.39640677,12.2310911 1,11.5586745 1,11.0086308 Z M16,11.0086308 C16,11.5586745 15.6035932,12.2310911 15.1275986,12.503088 L9.87240137,15.5060579 C9.3905874,15.7813801 9,15.5578743 9,15.0086308 L9,9.50051502 C9,8.95047139 9.39640677,8.27805477 9.87240137,8.00605786 L15.1275986,5.00308799 C15.6094126,4.72776572 16,4.95127159 16,5.50051502 L16,11.0086308 Z M9.34976149,6.98164278 C8.88045118,7.27044912 8.11286116,7.26633364 7.65023851,6.98164278 L2.84976149,4.02750307 C2.38045118,3.73869673 2.38713884,3.29611355 2.87661892,3.03254735 L7.62338108,0.476598498 C8.10752434,0.215905973 8.88713884,0.213032303 9.37661892,0.476598498 L14.1233811,3.03254735 C14.6075243,3.29323988 14.6128612,3.74281221 14.1502385,4.02750307 L9.34976149,6.98164278 Z" id="path-1"></path> \
+      </defs> \
+  </svg>'
+
+  const successMockedResponse = {
+    ok: true,
+    blob: jest.fn().mockResolvedValue(
+      new Blob([iconSvg], {
+        type: 'image/svg+xml'
+      })
+    )
+  }
+
+  const errorMockedResponse = {
+    ok: false,
+    statusText: 'Mocked response'
+  }
+
+  const badMimeTypeMockedResponse = {
+    ok: true,
+    blob: jest.fn().mockResolvedValue(
+      new Blob([iconSvg], {
+        type: 'text/plain'
+      })
+    )
+  }
+
+  beforeEach(() => {
+    jest.resetModules()
+    window.URL.createObjectURL = jest.fn().mockReturnValue('data://test-url')
+  })
+
+  it(`renders default icon when no Cozy Client`, () => {
+    const component = shallow(<AppIcon app={app} />).getElement()
+    expect(component).toMatchSnapshot()
+  })
+
+  it(`renders default when fetch error occurs`, () => {
+    context.client.link.client.fetch.mockImplementation(() => {
+      throw new Error('Mocked error')
+    })
```

Comment:
If the `fetchIcon` was passed as a property, it would be easier to mock and the tests would not have to know about the inner workings of the client.

IMHO there should be an

```
<DumpAppIcon />
class ClientAppIcon extends React.Component {
  constructor () {
    super(props, context)
    this.fetchIcon = this.fetchIcon.bind(this) 
  }

  fetchIcon () {
    const client = this.props.context
    // dealing with cozy.links, mime type etc... 
  }

  render () {
    return <DumbAppIcon fetchIcon={this.fetchIcon}/> 
  }
}
```